### PR TITLE
Planning algorithm fixes [AK-151]

### PR DESCRIPTION
**Pull request by karpinsk**
In progress - need to update according to the new start Time for floats

EDIT1: 
Algorithm works. Edge cases are handled, but feel free to stress-test 😄 . We are currently having problems with repeated tasks, so in db-seeder repeated tasks are simulated to create an example from our thesis.

All tests are currently happy-cases - need to create a proper error handling to update them, so bad cases will be handled in a separate task scope.

EDIT2: 
Tests are now really tests xD 
Also changed to @dimamik `newDate` function - thanks a lot!

### CODE REVIEW

File: apps/server/src/database/db-seeder.ts [hunk 1]
```diff
@@ -121,189 +118,213 @@ export const seedDatabase = async (app: INestApplication) => {
     repeatEvery: 1,
   }).save();
 
-  // Const tasks
+  // Repeated chunks added manually, because repeat doesn`t work now
```
Comment (dimamik) [inline apps/server/src/database/db-seeder.ts:121]: Why repeat is not working now?

File: apps/server/src/workers/taskPlanner.ts [hunk 2]
```diff
@@ -115,37 +111,18 @@ export class TaskPlanner {
         .delete()
         .from(Chunk)
         .where('taskId IN (:...ids)', { ids: floatTasks.map((t) => t.id) })
-        // Where start is after task.start
-        .andWhere('start >= :start', { start: task.chunkInfo.start })
+        // Where chunks are not done
+        .andWhere('isDone = :isDone', { isDone: false })
```
Comment (dimamik) [inline apps/server/src/workers/taskPlanner.ts:115]: We may also want to only delete chunks from the future.

File: apps/server/tests/tasks/planTask.e2e-spec.ts [hunk 3]
```diff
@@ -72,7 +72,169 @@ describe('PlanTask (e2e)', () => {
...
+        minChunkDuration: moment.duration(60, 'minutes'),
+        maxChunkDuration: moment.duration(60, 'minutes'),
+        deadline: newDate(new Date(2022, 12, 23, 0, 0)),
+        estimation: moment.duration(120, 'minutes'),
+        chillTime: moment.duration(0, 'minutes'),
+      },
+    }).save();
+
+    await taskPlanner.planTask(A);
+    await taskPlanner.planTask(B);
```
Comment (dimamik) [inline apps/server/tests/tasks/planTask.e2e-spec.ts:234]: Can we assert desired state of this task chunks? (That would be actually testing things, not just checking if planTask not throwing anything)

File: apps/server/tests/tasks/planTask.e2e-spec.ts [hunk 4]
```diff
@@ -244,92 +395,161 @@ describe('PlanTask (e2e)', () => {
...
+
+    const preparationCategory = await Category.create({
+      name: 'Preparation',
+      color: yellowColor,
+      user: user,
+    }).save();
+
+    // Now let's simulate already planned float task:
+
+    const A = await Task.create({
```
Comment (dimamik) [inline apps/server/tests/tasks/planTask.e2e-spec.ts:474]: I'm feeling like we could create a util function to not repeat those task creation processes.

File: apps/server/tests/tasks/planTask.e2e-spec.ts [hunk 5]
```diff
@@ -244,92 +395,161 @@ describe('PlanTask (e2e)', () => {
...
+      name: 'A',
       category: preparationCategory,
       priority: 'medium',
       isFloat: true,
       user: user,
       chunkInfo: {
-        start: newDate(new Date(2022, 11, 14)),
+        start: newDate(new Date(2029, 1, 3)),
```
Comment (dimamik) [inline apps/server/tests/tasks/planTask.e2e-spec.ts:481]: In one of my recent PRs I've introduced a newDate function that safely uses dates from the next week, nevertheless the test itself was run. Maybe we could merge main into that branch and use this function?

File: apps/server/tests/tasks/planTask.e2e-spec.ts [hunk 6]
```diff
@@ -244,92 +395,161 @@ describe('PlanTask (e2e)', () => {
...
-        estimation: moment.duration(5, 'hours'),
+        maxChunkDuration: moment.duration(1.5, 'hour'),
+        deadline: newDate(new Date(2029, 1, 5, 0, 0)),
+        estimation: moment.duration(20, 'hours'),
         chillTime: moment.duration(15, 'minutes'),
       },
     }).save();
 
-    // Create task breakdowns for the above task
+    await taskPlanner.planTask(A);
```
Comment (dimamik) [inline apps/server/tests/tasks/planTask.e2e-spec.ts:490]: Again, we need to assert something in there, for those to be tests.

**Review by dimamik — CHANGES_REQUESTED** [review]
I wasn't checking the logic itself yet. But can we please make tests to be tests and assert desired state of the chunks after planning the task?

File: apps/server/tests/tasks/planTask.e2e-spec.ts (hunk 5 above)
Comment (karpinsk) [reply apps/server/tests/tasks/planTask.e2e-spec.ts:481]: Sure, that would be nice. I`ve seen that newDate function, but as you said it's not merged yet.

File: apps/server/src/database/db-seeder.ts (hunk 1 above)
Comment (karpinsk) [reply apps/server/src/database/db-seeder.ts:121]: Not really sure :\
From the calendar view, we can see only my custom-added repeated chunks.

File: apps/server/src/workers/taskPlanner.ts (hunk 2 above)
Comment (karpinsk) [reply apps/server/src/workers/taskPlanner.ts:115]: Good point! But I need to test it better to make sure that nothing crashes with such an assumption.
Let's have it like that for now, I'll keep it in mind 💯

File: apps/server/tests/tasks/planTask.e2e-spec.ts (hunk 4 above)
Comment (karpinsk) [reply apps/server/tests/tasks/planTask.e2e-spec.ts:532]: Tried to create some util, but it looks like overengineering because every test has its own dataset, sharing a really small amount of common data.

Comment (dimamik) [reply apps/server/tests/tasks/planTask.e2e-spec.ts:532]: I meant inside the single test suite. Just something like createTask(...args), where args are all needed for creating the task args omitting defaults.

File: apps/server/src/workers/taskPlanner.ts (hunk 2 above)
Comment (dimamik) [reply apps/server/src/workers/taskPlanner.ts:115]: You can first write tests with assertions (for them to be real tests) and then change this part and check if everything is smooth still.

File: apps/server/src/database/db-seeder.ts (hunk 1 above)
Comment (dimamik) [reply apps/server/src/database/db-seeder.ts:121]: I'm not following.

The reason calendar is not rendering those, is because of the front. It has nothing to do with backend, so please, let's add repeat through normal taskService.createConst() method.

**Comment by dimamik** [comment]
And by the way, @karpinsk, can you please write your messages not in the PR's description, but in the comments like this one, because we will get notified then and will see a clean linear history of changes! Thanks in advance 💯

**Review by szarbartosz — APPROVED** [review]

**Review by dimamik — APPROVED** [review]
Great job!
